### PR TITLE
Add Pusher and Pubnub adaptor tutorials

### PR DESCRIPTION
**Pull request by SimonWoolf**
Pusher and Pubnub tutorials.

Decided to make the meat of the examples interoperability examples, so they all use both the Pusher and Ably or Pubnub and Ably client libs. Main reason is in the case of Pusher, that seems to be what trips people up (channel name translation).

The Pusher interop examples are all 'one-way' (js: ably->pusher, node and ruby: pusher->ably), as the Pusher node and ruby client libs are http-only and the browser lib can't publish without an auth server. The Pubnub interop examples are all 'two-way' - you can publish with either lib, and it receives with both.

### CODE REVIEW

File: content/tutorials/pubnub-adaptor.textile [hunk 1]
```diff
@@ -0,0 +1,199 @@
+---
+title: Pubnub Protocol Adaptor Tutorial
+section: tutorials
+index: 15
+---
+
+The Pubnub protocol adaptor allows you to interact with Ably using the Pubnub protocol.  You can use this to gradually migrate across from Pubnub to Ably. Rather than having to rewrite all your apps to use Ably client libraries at once, you can use the protocol adaptor, then migrate across to Ably client libraries one by one at your leisure, to take advantage of "Ably features not supported by the Pubnub client libraries":https://www.ably.io/features, such as "the ability to use websockets":https://support.ably.io/solution/articles/3000044831-which-transports-are-supported%2D (Pubnub client libraries support long-polling only). Alternatively, you can use a Pubnub client library on some platform for which no native Ably client library is available yet, such as Arduinos.
```
Comment (mattheworiordan) [inline content/tutorials/pubnub-adaptor.textile]: Platform is `Arduino`, not `Arduinos`, the devices  are `Arduinos`

File: content/tutorials/pubnub-adaptor.textile [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    :origin        => 'pubnub.ably.io',
+    :ssl           => true
+  )
+  ```
+
+  "See this step in Github":https://github.com/ably/tutorials/commit/c21bc0d
+
+Don't try to use different Ably API keys for the @publish_key@ and @subscribe_key@. Unlike Pubnub, Ably does not use different keys for publish and subscribe; instead, capabilities are connection-oriented, and the Adaptor will use whatever you pass as the @subscribe_key@ to create the Ably connection. If that key does not have publish capabilities, you will not be able to publish, whatever the @publish_key@ has.
+
+If we don't have an example in the language you'll be using, check the "Pubnub documentation for that library":https://pubnub.com/docs to find the equivalent options to the ones in the javascript example (for realtime libraries) or the ruby or nodejs example (for REST libraries) above.
```
Comment (mattheworiordan) [inline content/tutorials/pubnub-adaptor.textile]: `s/javascript/Javascript/`, `s/ruby/Ruby`, `s/nodejs/Node.js`

File: content/tutorials/pubnub-adaptor.textile [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+      puts "Pubnub client received a message: #{env.message.inspect}"
+    end
+
+    # Publish a message with the Ably client
+    # Note that we use null as the message name, since any
+    # message name will be ignored by the Pubnub client
+    ablyChannel.publish(nil, { some: 'data sent by the Ably client library' })
+
+    # Publish a message with the Pubnub client
+    pubnub.publish(channel: channel_name, message: { some: 'data sent by the Pubnub client library' }) {}
```
Comment (mattheworiordan) [inline content/tutorials/pubnub-adaptor.textile]: This line is too long for the code formatter

File: content/tutorials/pubnub-adaptor.textile (hunk 1 above)
Comment (mattheworiordan) [inline content/tutorials/pubnub-adaptor.textile]: Spacing `protocol.  You` -> `protocol. You`
Can you we change the sentence `You can use this to gradually migrate across from Pubnub to Ably. See some of the reasons why people [migrate from PubNub to Ably](https://www.ably.io/compare/ably-vs-pubnub)."

File: content/tutorials/pusher-adaptor.textile [hunk 4]
```diff
@@ -0,0 +1,206 @@
+---
+title: Pusher Protocol Adaptor Tutorial
+section: tutorials
+index: 15
+---
+
+The Pusher protocol adaptor allows you to interact with Ably using the Pusher protocol.  You can use this to gradually migrate across from Pusher to Ably. Rather than having to rewrite all your apps to use Ably client libraries at once, you can use the protocol adaptor, then migrate across to Ably client libraries one by one at your leisure, to take advantage of "Ably features not supported by the Pusher client libraries":https://www.ably.io/features such as "connection state recovery":https://support.ably.io/support/solutions/articles/3000044639-connection-state-recovery. Alternatively, you can use a Pusher client library on some platform for which no native Ably client library is available yet.
```
Comment (mattheworiordan) [inline content/tutorials/pusher-adaptor.textile]: Please apply the same comments from PubNub to here

File: content/tutorials/pubnub-adaptor.textile (hunk 1 above)
Comment (mattheworiordan) [inline content/tutorials/pubnub-adaptor.textile]: "Ably features not supported by the Pubnub client libraries" should link to https://www.ably.io/compare/ably-vs-pubnub

File: content/tutorials/pusher-adaptor.textile [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+Ably and Pusher both have various (incompatible) restrictions on channel names, and use namespaces very differently. So when using the adaptor, channel names are mapped. For example if you connect to the channel "foo" using the Pusher adaptor, it will actually use the Ably channel "public:foo"; and if you want to communicate with the Pusher client library from an Ably client library, it would need to connect to that channel.
+
+The translation rules are as follows:
+
+* Pusher public channels (no prefix) are mapped to the the Ably @public:@ namespace.
+* Pusher private channels (@private-@ prefix) are mapped to the the Ably @private:@ namespace.
+* Pusher presence channels (@presence-@ prefix) are mapped to the the Ably @presence:@ namespace.
+* Ably channels not in any of the @public:@, @private:@, or @presence:@ namespaces get a @private-ablyroot-@ prefix when seen from a Pusher client.
+* Semicolons are banned in Ably channel names; colons are banned in Pusher channel names. So the adaptor maps one to the other: semicolons in Pusher channel names become colons in Ably channel names.
```
Comment (mattheworiordan) [inline content/tutorials/pusher-adaptor.textile]: Semicolons are not banned in Ably channel names?  Channel `foo:bar:x` is valid?

File: content/tutorials/pusher-adaptor.textile [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+  We recommend that you clone the repo locally:
+
+  bc[sh]. git clone https://github.com/ably/tutorials.git
+
+  Checkout the tutorial branch:
+
+  bc[sh]. git checkout pusher-adaptor-javascript
+
+  And then run the demo locally by adding your "Ably API key":https://support.ably.io/support/solutions/articles/3000030502-setting-up-and-managing-api-keys to @example.html@ and opening the page in your browser.
```
Comment (mattheworiordan) [inline content/tutorials/pusher-adaptor.textile]: Should we not mention here that they set up the Pusher compatibility mode?

File: content/tutorials/pubnub-adaptor.textile [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+  Checkout the tutorial branch:
+
+  bc[sh]. git checkout pubnub-adaptor-ruby
+
+  And then run the demo locally by adding your "Ably API key":https://support.ably.io/support/solutions/articles/3000030502-setting-up-and-managing-api-keys to @example.rb@ and running the demo with @bundle exec ruby example.rb@
+
+h2. Next steps
+
+# Full usage notes for the Pubnub adaptor are available at: "Using the Ably Pubnub Protocol Adaptor":https://support.ably.io/support/solutions/articles/3000055086-using-the-ably-pusher-protocol-adaptor/.
```
Comment (mattheworiordan) [inline content/tutorials/pubnub-adaptor.textile]: Can we add the first bullet "[Review the many reasons to migrate from PubNub to Ably](https://www.ably.io/compare/ably-vs-pubnub)"

File: content/tutorials/pusher-adaptor.textile [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+  Checkout the tutorial branch:
+
+  bc[sh]. git checkout pusher-adaptor-ruby
+
+  And then run the demo locally by adding your "Ably API key":https://support.ably.io/support/solutions/articles/3000030502-setting-up-and-managing-api-keys to @example.rb@ and running the demo with @bundle exec ruby example.rb@
+
+h2. Next steps
+
+# Full usage notes for the Pusher adaptor are available at: "Using the Ably Pusher Protocol Adaptor":https://support.ably.io/support/solutions/articles/3000055086-using-the-ably-pusher-protocol-adaptor/.
```
Comment (mattheworiordan) [inline content/tutorials/pusher-adaptor.textile]: Can we add the first bullet "[Review the many reasons to migrate from Pusher to Ably](https://www.ably.io/compare/ably-vs-pusher)"

File: content/tutorials/pusher-adaptor.textile (hunk 6 above)
Comment (SimonWoolf) [inline content/tutorials/pusher-adaptor.textile]: Sure, could mention it - was assuming that they'd been following the tutorial, in which case they'd have done that step

**Comment by SimonWoolf** [comment]
Made requested changes: 9fc1680

(Not sure if you want to bother reviewing the code examples; if so, handy links:
https://github.com/ably/tutorials/blob/pubnub-adaptor-javascript/example.html
https://github.com/ably/tutorials/blob/pubnub-adaptor-ruby/example.rb
https://github.com/ably/tutorials/blob/pubnub-adaptor-nodejs/example.js
https://github.com/ably/tutorials/blob/pusher-adaptor-javascript/example.html
https://github.com/ably/tutorials/blob/pusher-adaptor-ruby/example.rb
https://github.com/ably/tutorials/blob/pusher-adaptor-nodejs/example.js
)

**Comment by SimonWoolf** [comment]
Added 78853ce and ca2d02d to rename adaptor->adapter per https://github.com/ably/docs/pull/209 (as that PR wasn't done on a branch with the adapter tutorials on it)

**Comment by mattheworiordan** [comment]
Thanks @SimonWoolf LGTM

**Pull request merged**